File: backend/app/routers/yields.py
"""Product Yields — harvest logging. A domain separate from PO/QBO sales; see
services/yields.py. No router-level role floor: 'field' (the harvest kiosk
role) is rank 0, the floor of the whole app, so bare `current_user` is already
correct for reads/submissions. Product management stays `require_editor`,
matching the rest of the app's catalog-editing endpoints."""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from ..auth import AuthedUser, app_role, current_user, require_admin, require_editor
from ..reused_db import reused_conn
from ..schemas import PageResponse
from ..services import yields as yields_svc

router = APIRouter(prefix="/api/yields", tags=["yields"])


def _actor(user: AuthedUser) -> str | None:
    return user.email or user.id


# --- products ----------------------------------------------------------------


class ProductIn(BaseModel):
    name: str
    notes: str | None = None
    lot_code_prefix: str | None = None


class ProductPatch(BaseModel):
    name: str | None = None
    active: bool | None = None
    notes: str | None = None
    lot_code_prefix: str | None = None


@router.get("/products")
def list_products(include_inactive: bool = False, _: AuthedUser = Depends(current_user)) -> list[dict]:
    with reused_conn() as conn:
        return yields_svc.list_products(conn, include_inactive=include_inactive)


@router.post("/products")
def create_product(body: ProductIn, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        return yields_svc.create_product(
            conn, body.name, body.notes, lot_code_prefix=body.lot_code_prefix, actor=_actor(user),
        )


@router.post("/products/{product_id}")
def update_product(product_id: int, body: ProductPatch, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        return yields_svc.update_product(
            conn, product_id, name=body.name, active=body.active, notes=body.notes,
            lot_code_prefix=body.lot_code_prefix, actor=_actor(user),
        )


@router.delete("/products/{product_id}")
def delete_product(product_id: int, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        yields_svc.delete_product(conn, product_id, actor=_actor(user))
    return {"ok": True}


# --- yield product <-> sales SKU links ----------------------------------------


class LinkIn(BaseModel):
    yield_product_id: int
    sales_product_name: str


@router.get("/links")
def list_links(yield_product_id: int | None = None, _: AuthedUser = Depends(require_editor)) -> list[dict]:
    with reused_conn() as conn:
        return yields_svc.list_links(conn, yield_product_id=yield_product_id)


@router.post("/links")
def create_link(body: LinkIn, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        return yields_svc.create_link(
            conn, body.yield_product_id, body.sales_product_name, actor=_actor(user),
        )


@router.delete("/links/{link_id}")
def delete_link(link_id: int, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        yields_svc.delete_link(conn, link_id, actor=_actor(user))
    return {"ok": True}


@router.get("/sales-product-names")
def sales_product_names(_: AuthedUser = Depends(require_editor)) -> list[str]:
    with reused_conn() as conn:
        return yields_svc.sales_product_names(conn)


# --- entries -----------------------------------------------------------------


class EntryIn(BaseModel):
    yield_product_id: int
    harvest_date: date
    weight: float
    unit: str = "oz"
    tray_count: int = 0
    discarded_tray_count: int = 0
    lot_code: str | None = None
    harvested_by: str
    notes: str | None = None


class EntryPatch(BaseModel):
    weight: float | None = None
    unit: str | None = None
    tray_count: int | None = None
    discarded_tray_count: int | None = None
    lot_code: str | None = None
    harvested_by: str | None = None
    notes: str | None = None
    harvest_date: date | None = None


class VoidIn(BaseModel):
    reason: str | None = None


@router.post("/entries")
def create_entry(body: EntryIn, user: AuthedUser = Depends(current_user)) -> dict:
    with reused_conn() as conn:
        return yields_svc.create_entry(
            conn,
            yield_product_id=body.yield_product_id,
            harvest_date=body.harvest_date,
            weight=body.weight,
            unit=body.unit,
            tray_count=body.tray_count,
            discarded_tray_count=body.discarded_tray_count,
            lot_code=body.lot_code,
            harvested_by=body.harvested_by,
            notes=body.notes,
            submitted_by=_actor(user) or "",
        )


class EntryImportRow(BaseModel):
    yield_product_id: int
    harvest_date: date
    weight: float
    unit: str = "oz"
    lot_code: str | None = None
    harvested_by: str


class EntriesImportIn(BaseModel):
    entries: list[EntryImportRow]


# Admin-only, and deliberately not shared with the plain current_user floor
# every other entries endpoint uses — the kiosk's 'field' role has no
# legitimate reason to bulk-import historical data, unlike submitting its
# own entries one at a time.
@router.post("/entries/import")
def import_entries(body: EntriesImportIn, user: AuthedUser = Depends(require_admin)) -> dict:
    if not body.entries:
        raise HTTPException(422, "no entries to import")
    with reused_conn() as conn:
        n = yields_svc.import_entries(
            conn, [e.model_dump() for e in body.entries], actor=_actor(user) or "",
        )
    return {"ok": True, "created": n}


@router.get("/entries")
def list_entries(
    yield_product_id: int | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    harvested_by: str | None = None,
    submitted_by: str | None = None,
    include_voided: bool = False,
    _: AuthedUser = Depends(current_user),
) -> list[dict]:
    with reused_conn() as conn:
        return yields_svc.list_entries(
            conn,
            yield_product_id=yield_product_id,
            date_from=date_from,
            date_to=date_to,
            harvested_by=harvested_by,
            submitted_by=submitted_by,
            include_voided=include_voided,
        )


@router.post("/entries/{entry_id}")
def update_entry(entry_id: int, body: EntryPatch, user: AuthedUser = Depends(current_user)) -> dict:
    with reused_conn() as conn:
        return yields_svc.update_entry(
            conn, entry_id, body.model_dump(exclude_unset=True),
            actor=_actor(user), actor_role=app_role(user.email),
        )


@router.post("/entries/{entry_id}/void")
def void_entry(entry_id: int, body: VoidIn, user: AuthedUser = Depends(current_user)) -> dict:
    with reused_conn() as conn:
        return yields_svc.void_entry(
            conn, entry_id, body.reason, actor=_actor(user), actor_role=app_role(user.email),
        )


# --- trends ------------------------------------------------------------------


@router.get("/trends")
def trends(
    date_from: date | None = None,
    date_to: date | None = None,
    yield_product_id: list[int] | None = Query(None),
    grain: str = "month",
    _: AuthedUser = Depends(current_user),
) -> PageResponse:
    if grain not in yields_svc.GRAINS:
        raise HTTPException(422, f"grain must be one of {', '.join(yields_svc.GRAINS)}")
    with reused_conn() as conn:
        return yields_svc.trends(
            conn, date_from=date_from, date_to=date_to,
            yield_product_ids=yield_product_id, grain=grain,
        )


# --- notes -------------------------------------------------------------------
# Admin-only for now (per the user, while both role sides are still being
# tested) — tighten/loosen later by swapping require_admin below.


class NoteIn(BaseModel):
    lot_code: str | None = None
    note: str
    note_date: date | None = None


@router.get("/notes")
def list_notes(lot_code: str | None = None, _: AuthedUser = Depends(require_admin)) -> list[dict]:
    with reused_conn() as conn:
        return yields_svc.list_notes(conn, lot_code=lot_code)


@router.post("/notes")
def create_note(body: NoteIn, user: AuthedUser = Depends(require_admin)) -> dict:
    with reused_conn() as conn:
        return yields_svc.create_note(
            conn,
            lot_code=body.lot_code,
            note=body.note,
            note_date=body.note_date,
            submitted_by=_actor(user) or "",
        )


@router.delete("/notes/{note_id}")
def delete_note(note_id: int, user: AuthedUser = Depends(require_admin)) -> dict:
    with reused_conn() as conn:
        yields_svc.delete_note(conn, note_id, actor=_actor(user))
    return {"ok": True}


# --- employees -----------------------------------------------------------------


class EmployeeIn(BaseModel):
    name: str


class EmployeePatch(BaseModel):
    name: str | None = None
    active: bool | None = None


@router.get("/employees")
def list_employees(include_inactive: bool = False, _: AuthedUser = Depends(current_user)) -> list[dict]:
    with reused_conn() as conn:
        return yields_svc.list_employees(conn, include_inactive=include_inactive)


@router.post("/employees")
def create_employee(body: EmployeeIn, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        return yields_svc.create_employee(conn, body.name, actor=_actor(user))


@router.post("/employees/{employee_id}")
def update_employee(employee_id: int, body: EmployeePatch, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        return yields_svc.update_employee(
            conn, employee_id, name=body.name, active=body.active, actor=_actor(user),
        )


@router.delete("/employees/{employee_id}")
def delete_employee(employee_id: int, user: AuthedUser = Depends(require_editor)) -> dict:
    with reused_conn() as conn:
        yields_svc.delete_employee(conn, employee_id, actor=_actor(user))
    return {"ok": True}

File: web/src/api/yields.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiGet, apiSend } from "@/lib/api";
import type { PageResponse } from "@/api/schema";

export type YieldUnit = "oz" | "lb" | "g";
export type YieldGrain = "week" | "month" | "quarter" | "year";

export interface YieldProduct {
  id: number;
  name: string;
  active: boolean;
  notes: string | null;
  /** Short lot-code prefix (e.g. "TK") — the entry form auto-fills the lot
   *  code field with this for traceability, without retyping it every time. */
  lot_code_prefix: string | null;
  /** Computed, not a DB column — true if any of this product's entries has
   *  its own free-text notes, or shares a lot_code with a yield_notes
   *  grower observation. Not related to `notes` above (a catalog
   *  description field on the product itself). */
  has_notes: boolean;
}

export interface YieldEntry {
  id: number;
  yield_product_id: number;
  product_name: string;
  /** The product's lot-code prefix (e.g. "TK") — only populated on list
   *  results (joined), not on the row returned from create/update/void. */
  product_lot_code_prefix: string | null;
  harvest_date: string;
  weight: number;
  unit: YieldUnit;
  tray_count: number;
  discarded_tray_count: number;
  lot_code: string | null;
  harvested_by: string;
  submitted_by: string;
  notes: string | null;
  voided: boolean;
  void_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface YieldEntryFilters {
  yield_product_id?: number;
  date_from?: string;
  date_to?: string;
  harvested_by?: string;
  submitted_by?: string;
  include_voided?: boolean;
  /** Only entries with a non-empty own `notes` field — for the Notes
   *  page's merged feed. */
  has_notes?: boolean;
}

export interface YieldEntryIn {
  yield_product_id: number;
  harvest_date: string;
  weight: number;
  unit: YieldUnit;
  tray_count: number;
  discarded_tray_count: number;
  lot_code?: string | null;
  harvested_by: string;
  notes?: string | null;
}

export function useYieldProducts(includeInactive = false) {
  return useQuery({
    queryKey: ["yield-products", includeInactive],
    queryFn: () => apiGet<YieldProduct[]>("/api/yields/products", { include_inactive: includeInactive }),
    staleTime: 60_000,
  });
}

// create/update's RETURNING clause is yield_products alone — has_notes is
// computed only by list_products' two-query join, so it's never on these.
type YieldProductMutationResult = Omit<YieldProduct, "has_notes">;

export function useCreateYieldProduct() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: { name: string; notes?: string | null; lot_code_prefix?: string | null }) =>
      apiSend<YieldProductMutationResult>("POST", "/api/yields/products", body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-products"] }),
  });
}

export function useUpdateYieldProduct() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      id,
      ...body
    }: {
      id: number;
      name?: string;
      active?: boolean;
      notes?: string | null;
      lot_code_prefix?: string | null;
    }) => apiSend<YieldProductMutationResult>("POST", `/api/yields/products/${id}`, body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-products"] }),
  });
}

/** Only succeeds while the product has no harvest entries yet (backend 409s
 *  with code "in_use" otherwise) — retire it via useUpdateYieldProduct instead. */
export function useDeleteYieldProduct() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => apiSend<{ ok: boolean }>("DELETE", `/api/yields/products/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-products"] }),
  });
}

export interface YieldLink {
  id: number;
  yield_product_id: number;
  product_name: string;
  sales_product_name: string;
}

export function useYieldLinks(yieldProductId?: number) {
  return useQuery({
    queryKey: ["yield-links", yieldProductId],
    queryFn: () =>
      apiGet<YieldLink[]>("/api/yields/links", yieldProductId ? { yield_product_id: yieldProductId } : undefined),
    staleTime: 30_000,
  });
}

export function useYieldSalesProductNames() {
  return useQuery({
    queryKey: ["yield-sales-product-names"],
    queryFn: () => apiGet<string[]>("/api/yields/sales-product-names"),
    staleTime: 60_000,
  });
}

export function useCreateYieldLink() {
  const qc = useQueryClient();
  return useMutation({
    // The create endpoint's RETURNING clause has no product_name (that's a
    // join only list_links does) — Omit it rather than claiming YieldLink's
    // full shape for a response that doesn't have it.
    mutationFn: (body: { yield_product_id: number; sales_product_name: string }) =>
      apiSend<Omit<YieldLink, "product_name">>("POST", "/api/yields/links", body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-links"] }),
  });
}

export function useDeleteYieldLink() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => apiSend<{ ok: boolean }>("DELETE", `/api/yields/links/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-links"] }),
  });
}

export function useYieldEntries(filters: YieldEntryFilters = {}, options?: { enabled?: boolean }) {
  return useQuery({
    queryKey: ["yield-entries", filters],
    queryFn: () => apiGet<YieldEntry[]>("/api/yields/entries", { ...filters }),
    staleTime: 15_000,
    enabled: options?.enabled,
  });
}

// The backend's RETURNING clause for create/update/void is `yield_entries`
// alone (no join to yield_products), so these three never carry
// product_name/product_lot_code_prefix — only list_entries's joined rows do.
type YieldEntryMutationResult = Omit<YieldEntry, "product_name" | "product_lot_code_prefix">;

export function useCreateYieldEntry() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: YieldEntryIn) =>
      apiSend<YieldEntryMutationResult>("POST", "/api/yields/entries", body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-entries"] }),
  });
}

export interface YieldEntryImportRow {
  yield_product_id: number;
  harvest_date: string;
  weight: number;
  unit: YieldUnit;
  tray_count?: number;
  lot_code?: string | null;
  harvested_by: string;
}

/** Admin-only, one-transaction bulk insert for historical data (the CSV
 *  import page) — tray_count is optional per row (defaults to 0; most
 *  historical records don't carry it), discarded_tray_count/notes are
 *  never set, and a single audit_log row summarizes the whole batch. The
 *  backend skips (doesn't create) any row that already matches a non-voided
 *  entry for the same product + date + weight + unit (also + lot_code when
 *  the row has one) — reported back as skipped_duplicates, so re-importing
 *  the template's own example rows, or the same file twice, doesn't double
 *  up, while two different entries that happen to share a lot code (e.g.
 *  two people harvesting into the same labeled bin) still both get created. */
export function useImportYieldEntries() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (entries: YieldEntryImportRow[]) =>
      apiSend<{ ok: boolean; created: number; skipped_duplicates: number }>(
        "POST",
        "/api/yields/entries/import",
        { entries },
      ),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-entries"] }),
  });
}

export interface YieldEntryPatch {
  harvest_date?: string;
  weight?: number;
  unit?: YieldUnit;
  tray_count?: number;
  discarded_tray_count?: number;
  lot_code?: string | null;
  harvested_by?: string;
  notes?: string | null;
}

/** Corrections to an already-logged entry. Backend enforces: a 'field'-rank
 *  caller may only touch its own same-day entries; every other role
 *  (viewer+) is unrestricted — see services/yields.py's _assert_can_touch. */
export function useUpdateYieldEntry() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, ...body }: YieldEntryPatch & { id: number }) =>
      apiSend<YieldEntryMutationResult>("POST", `/api/yields/entries/${id}`, body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-entries"] }),
  });
}

export function useVoidYieldEntry() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, reason }: { id: number; reason?: string | null }) =>
      apiSend<YieldEntryMutationResult>("POST", `/api/yields/entries/${id}/void`, { reason }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-entries"] }),
  });
}

export interface YieldTrendsFilters {
  date_from?: string;
  date_to?: string;
  yield_product_id?: number[];
  grain?: YieldGrain;
}

export function useYieldTrends(filters: YieldTrendsFilters) {
  return useQuery({
    queryKey: ["yield-trends", filters],
    queryFn: () => apiGet<PageResponse>("/api/yields/trends", { ...filters }),
    staleTime: 30_000,
  });
}

export interface YieldNote {
  id: number;
  /** The lot it's about (the default/primary case), or null for a general
   *  note not specific to any one lot. */
  lot_code: string | null;
  note_date: string;
  note: string;
  submitted_by: string;
  created_at: string;
}

/** Admin-only for now, both here and on the backend (require_admin) — see
 *  routers/yields.py's notes section. */
export function useYieldNotes(lotCode?: string) {
  return useQuery({
    queryKey: ["yield-notes", lotCode],
    queryFn: () => apiGet<YieldNote[]>("/api/yields/notes", lotCode ? { lot_code: lotCode } : undefined),
    staleTime: 30_000,
  });
}

export function useCreateYieldNote() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: { lot_code?: string | null; note: string; note_date?: string | null }) =>
      apiSend<YieldNote>("POST", "/api/yields/notes", body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-notes"] }),
  });
}

export function useDeleteYieldNote() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => apiSend<{ ok: boolean }>("DELETE", `/api/yields/notes/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-notes"] }),
  });
}

export interface YieldEmployee {
  id: number;
  name: string;
  active: boolean;
}

/** The kiosk's "harvested by" roster — a curated suggestion list, not a FK:
 *  yield_entries.harvested_by stays free text either way. */
export function useYieldEmployees(includeInactive = false) {
  return useQuery({
    queryKey: ["yield-employees", includeInactive],
    queryFn: () => apiGet<YieldEmployee[]>("/api/yields/employees", { include_inactive: includeInactive }),
    staleTime: 60_000,
  });
}

export function useCreateYieldEmployee() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: { name: string }) => apiSend<YieldEmployee>("POST", "/api/yields/employees", body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-employees"] }),
  });
}

export function useUpdateYieldEmployee() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, ...body }: { id: number; name?: string; active?: boolean }) =>
      apiSend<YieldEmployee>("POST", `/api/yields/employees/${id}`, body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-employees"] }),
  });
}

export function useDeleteYieldEmployee() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => apiSend<{ ok: boolean }>("DELETE", `/api/yields/employees/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["yield-employees"] }),
  });
}
